onnx/defs/math/defs.cc

```diff
@@ -2006,6 +2007,273 @@ ONNX_OPERATOR_SET_SCHEMA(
+  If `zero_points` is not provided then zero_points will be set to `2^(bits - 1)`.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    MatMulNBits,
+    23,
+    OpSchema()
+        .SetDoc(MatMulNBits_ver23_doc)
+        .Attr(
+            "K",
+            "Size of each input feature. If not present, should be inferred from input A.",
+            AttributeProto::INT,
+            OPTIONAL_VALUE)
+        .Attr(
+            "N",
+            "Size of each output feature. If not present, should be inferred from input B.",
+            AttributeProto::INT,
+            OPTIONAL_VALUE)
+        .Attr(
+            "accuracy_level",
+            "The minimum accuracy level of input A, can be: "
+            "0(unset), 1(float), 2(float16), 3(Bfloat16), 4(int8) (default unset). It is used to control how input A "
+            "is quantized or downcast internally while doing computation, for example: 0 means input A will not be "
+            "quantized or downcast while doing computation, 4 means input A can be quantized with the same block_size"
+            "to int8 internally from type T1.",
+            AttributeProto::INT,
+            static_cast<int64_t>(0))
+        .Attr(
+            "bits",
+            "Number of bits used for weight quantization (default 4)",
+            AttributeProto::INT,
+            static_cast<int64_t>(4))
+        .Attr(
+            "block_size",
+            "Number of groupsize used for weight quantization (default 128)."
+            "It needs to be a power of 2 and not smaller than 16",
+            AttributeProto::INT,
+            static_cast<int64_t>(128))
+        .Input(0, "A", "The input tensor, not quantized", "T1", OpSchema::Single, true, 1, OpSchema::NonDifferentiable)
+        .Input(1, "B", "1 or 2 dimensional data blob", "T2", OpSchema::Single, true, 1, OpSchema::NonDifferentiable)
+        .Input(2, "scales", "quantization scales", "T1", OpSchema::Single, true, 1, OpSchema::NonDifferentiable)
+        .Input(
+            3,
+            "zero_points",
+            "quantization zero points",
+            "T3",
+            OpSchema::Optional,
+            true,
+            1,
+            OpSchema::NonDifferentiable)
+        .Input(
+            4,
+            "bias",
+            "Bias to add to result. It should have shape [N]",
+            "T1",
+            OpSchema::Optional,
+            true,
+            1,
+            OpSchema::NonDifferentiable)
+        .Output(
+            0,
+            "Y",
+            "Matrix multiply results. The output tensor has the same rank as the input.",
+            "T1",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::NonDifferentiable)
+        .TypeConstraint("T1", {"tensor(float)", "tensor(float16)"},
+                        "Constrain input and output type to float/half_float tensors.")
+        .TypeConstraint("T2", {"tensor(uint8)"}, "Constrain quantized weight types to uint8/int32.")
```

Comment:
Is the `int32` a typo? Same question 2 lines below.

Reply:
Good catch. Yes, both are typos from an earlier draft of the code that was missed I will remove.